File: 数据结构/3.4/No1190/main.py
class Solution:
    def reverseParentheses(self, s: str) -> str:
        st = []
        n = len(s)
        for i in range(n):
            if len(st) > 0 and s[i] == ")":
                s[st[-1] + 1 : i] = s[i - 1 : st[-1] : -1]
            elif s[i] == "(":
                st.append(i)


def main():
    s = "(ed(et(oc))el)"
    sol = Solution()
    print(sol.reverseParentheses(s))


main()

File: 数据结构/3.4/No1190/main.cpp
#include <algorithm>
#include <iostream>
#include <stack>
#include <string>
using namespace std;

class Solution {
public:
  string reverseParentheses(string s) {
    stack<string> st;
    string res;
    for (auto &ch : s) {
      if (ch == '(') {
        st.push(res);
        res = "";
      } else if (ch == ')') {
        reverse(res.begin(), res.end());
        res = st.top() + res;
        st.pop();
      } else {
        res.push_back(ch);
      }
    }
    return res;
  }
};

signed main() {
  string s = "(ed(et(oc))el)";
  Solution sol = Solution();
  cout << sol.reverseParentheses(s) << endl;

  return 0;
}